server/sonar-server/src/main/java/org/sonar/server/component/index/ComponentIndex.java

```diff
@@ -0,0 +1,105 @@
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.index.query.QueryBuilders;
+import org.elasticsearch.search.SearchHit;
+import org.sonar.core.util.stream.Collectors;
+import org.sonar.server.es.BaseIndex;
+import org.sonar.server.es.DefaultIndexSettings;
+import org.sonar.server.es.EsClient;
+import org.sonar.server.user.UserSession;
+
+import static org.elasticsearch.index.query.QueryBuilders.boolQuery;
+import static org.elasticsearch.index.query.QueryBuilders.matchAllQuery;
+import static org.elasticsearch.index.query.QueryBuilders.matchQuery;
+import static org.elasticsearch.index.query.QueryBuilders.termQuery;
+import static org.sonar.server.component.index.ComponentIndexDefinition.FIELD_AUTHORIZATION_GROUPS;
+import static org.sonar.server.component.index.ComponentIndexDefinition.FIELD_AUTHORIZATION_USERS;
+import static org.sonar.server.component.index.ComponentIndexDefinition.FIELD_KEY;
+import static org.sonar.server.component.index.ComponentIndexDefinition.FIELD_NAME;
+import static org.sonar.server.component.index.ComponentIndexDefinition.FIELD_QUALIFIER;
+import static org.sonar.server.component.index.ComponentIndexDefinition.INDEX_COMPONENTS;
+import static org.sonar.server.component.index.ComponentIndexDefinition.TYPE_AUTHORIZATION;
+import static org.sonar.server.component.index.ComponentIndexDefinition.TYPE_COMPONENT;
+
+public class ComponentIndex extends BaseIndex {
+
+  private final UserSession userSession;
+
+  public ComponentIndex(EsClient client, UserSession userSession) {
+    super(client);
+    this.userSession = userSession;
+  }
+
+  public List<String> search(ComponentIndexQuery query) {
+    SearchRequestBuilder request = getClient()
+      .prepareSearch(INDEX_COMPONENTS)
+      .setTypes(TYPE_COMPONENT)
+      .setFetchSource(false);
+
+    query.getLimit().ifPresent(request::setSize);
+
+    request.setQuery(createQuery(query));
+
+    return Arrays.stream(request.get().getHits().hits())
+      .map(SearchHit::getId)
+      .collect(Collectors.toList());
+  }
+
+  private QueryBuilder createQuery(ComponentIndexQuery query) {
+    BoolQueryBuilder esQuery = boolQuery();
+    esQuery.filter(createAuthorizationFilter());
+
+    query.getQualifier().ifPresent(q -> esQuery.filter(termQuery(FIELD_QUALIFIER, q)));
+
+    // We will truncate the search to the maximum length of nGrams in the index.
+    // Otherwise the search would for sure not find any results.
+    String truncatedQuery = StringUtils.left(query.getQuery(), DefaultIndexSettings.MAXIMUM_NGRAM_LENGTH);
+
+    return esQuery.must(boolQuery()
+      .should(matchQuery(FIELD_NAME + "." + SEARCH_PARTIAL_SUFFIX, truncatedQuery))
+      .should(matchQuery(FIELD_KEY + "." + SORT_SUFFIX, query.getQuery()).boost(3f)));
+  }
+
+  private QueryBuilder createAuthorizationFilter() {
+    Integer userLogin = userSession.getUserId();
+    Set<String> userGroupNames = userSession.getUserGroups();
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this use of "getUserGroups"; it is deprecated. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)